packages/tenant-process/src/services/tenantService.ts

```diff
@@ -1153,16 +1153,112 @@ export function tenantServiceBuilder(
       logger.info(`Retrieving Tenant with Selfcare Id ${selfcareId}`);
       return readModelService.getTenantBySelfcareId(selfcareId);
     },
+
+    async internalUpsertTenant(
+      internalTenantSeed: tenantApi.InternalTenantSeed,
+      { correlationId, logger }: WithLogger<AppContext>
+    ): Promise<Tenant> {
+      logger.info(
+        `Updating tenant with external id ${internalTenantSeed.externalId} via internal request`
+      );
+
+      const existingTenant = await readModelService.getTenantByExternalId(
+        internalTenantSeed.externalId
+      );
+
+      assertTenantExists(
+        unsafeBrandId(
+          `${internalTenantSeed.externalId.origin}/${internalTenantSeed.externalId.value}`
+        ),
+        existingTenant
+      );
+
+      const attributesExternalIds = internalTenantSeed.certifiedAttributes.map(
+        (externalId) =>
+          ExternalId.parse({
+            value: externalId.code,
+            origin: externalId.origin,
+          })
+      );
+
+      const attributesByExternalId =
+        await readModelService.getAttributesByExternalIds(
+          attributesExternalIds
+        );
+
+      for (const attributeToAssign of attributesExternalIds) {
+        if (
+          !attributesByExternalId.find(
+            (a) =>
+              a?.origin === attributeToAssign.origin &&
+              a?.code === attributeToAssign.value
+          )
+        ) {
+          throw attributeNotFound(
+            `${attributeToAssign.origin}/${attributeToAssign.value}`
+          );
+        }
+      }
+
+      const tenantWithNewAttributes = attributesByExternalId.reduce(
+        (acc: Tenant, attribute: Attribute) =>
+          assignCertifiedAttribute({
+            targetTenant: acc,
+            attribute,
+          }),
+        {
+          ...existingTenant.data,
+          updatedAt: new Date(),
+        }
+      );
+
+      const tenantCertifiedAttributesAssignedEvent =
+        toCreateEventTenantOnboardDetailsUpdated(
```

Comment:
if I'm not missing something
```suggestion
        TenantCertifiedAttributeAssigned(
```

Reply:
point is we always need to persist the event that is related to the change done.
If an unrelated event is persisted, consumers will not be able to apply proper logics.

In this case yes, let's persist a `TenantCertifiedAttributeAssigned` for each assigned attribute.
Tip: remember to persist all of them in the same transaction